Add introduction to the user manual (#2112)
kframework/k: USER_MANUAL.md

K User Manual
=============

**NOTE:** The K User Manual is still **under construction**; some features of K
may have partial or missing documentation.

Introduction
------------

### Why K?

The K Framework is a programming language and system design toolkit made for
practioners and researchers alike.

**K For Practioners:**
*K is a framework for deriving programming languages tools from their semantic
specifications.*

Typically, programming language tool development follows a similar pattern.
After a new programming language is designed, separate teams will develop
separate language tools (e.g. a compiler, interpreter, parser, symbolic
execution engine, etc). Code reuse is uncommon. The end result is that for each
new language, the same basic tools and patterns are re-implemented again and
again.

> K approaches the problem differently, by imagining how different tools could be
> derived generically from a common specification. In other words, for any
> language, a language-generic tool is a function that takes a language spec as
> input and produces a tool as output. The end result is that the exercise of
> designing new languages and their associated tooling is now reduced to
> developing a single language specfication from which we derive our tooling *for
> free*.

Reviewer: It's a little confusing that the tools are both "derived generically from a common specification" and that they take the "language spec as input".

```suggestion
K approaches the problem differently -- it generates each of these tools from a single language specification.
The work of programming language design and tool implementation are made separate concerns.
The end result is that the exercise of
designing new languages and their associated tooling is now reduced to
developing a single language specification from which we derive our tooling *for
free*.
```

    -   **Backus-Naur Form (BNF) grammar declarations:** define the
        operators that inhabit our primitive datatypes
    -   **lexical syntax declarations:** define lexemes/tokens for the
        lexer/tokenizer
    -   **syntax associativity declarations:** specify the
        associativity/grouping of our declared operators
    -   **syntax priority declarations:** specify the priority of
        potential ambiguous operators

2.  sentences that define our *system's state*, including:

    -   **configuration declarations:** define labelled, hierarchical records
        using an nested XML-like syntax

3.  sentences that define our *system's behavior*, including:

    -   **context declarations:** describe how primitives and configurations
        can simplify
    -   **context alias declarations:** define templates that can generate new
        contexts
    -   **rule declarations:** define how the system transitions from one state
        to the next

### K Process Overview

We now examine how the K tools are generally used. At a high-level, each K tool
can be understood as a blackbox with the following inputs and outputs:

```

>  K Compilation Process
> +============================================================+
> |                     +---------+                            |
> |  K Specification ---| kompile |--> Kore Specification --+  |
> |                     +---------+                         |  |
> +=========================================================|==+
>                                                           |
>  K Execution Process                                      |
> +=========================================================|==+
> |                                                         |  |
> |             +-------------------------------------------+  |
> |             |                                              |
> |             |       +---------+                            |
> |  K Term ----+-------| kparse  |--> K Term                  |
> |             |       +---------+                            |
> |             |                                              |
> |             |       +---------+                            |
> |  K Term ----+-------|  krun   |--> K Term                  |
> |             |       +---------+                            |
> |             |                                              |
> |             |       +---------+                            |
> |  K Claims --+-------| kprove  |--> K Claims                |
> |                     +---------+                            |
> |                                                            |
> +============================================================+

Reviewer: Kore has not been introduced yet, also `kparse` and `krun` take Programs as input.

```suggestion
 K Compilation Process
+============================================================+
|                     +---------+                            |
|  K Specification ---| kompile |--> Formal Specification +  |
|                     +---------+                         |  |
+=========================================================|==+
                                                          |
 K Execution Process                                      |
+=========================================================|==+
|                                                         |  |
|             +-------------------------------------------+  |
|             |                                              |
|             |       +---------+                            |
|  Program ---+-------| kparse  |--> K Term                  |
|             |       +---------+                            |
|             |                                              |
|             |       +---------+                            |
|  Program ---+-------|  krun   |--> K Term                  |
|             |       +---------+                            |
|             |                                              |
|             |       +---------+                            |
|  K Claims --+-------| kprove  |--> K Claims                |
|                     +---------+                            |
|                                                            |
+============================================================+
```

```

where:

-   process outlines are denoted by boxes with double-lined borders

> -   programs are denoted by boxes with single-lined borders

Reviewer: I think we can avoid confusion between programs in the programming language, and executables.

```suggestion
-   executables are denoted by boxes with single-lined borders
```

K User Manual
=============

**NOTE:** The K User Manual is still **under construction**; some features of K
may have partial or missing documentation.

Introduction
------------

### Why K?

The K Framework is a programming language and system design toolkit made for
practioners and researchers alike.

**K For Practioners:**
*K is a framework for deriving programming languages tools from their semantic
specifications.*

Typically, programming language tools are developed as follows: for each

> language *L*, a separate team develops tool *T* (e.g. a compiler, interpreter,

Reviewer 2: I don't know that we want to use this type of capital letter style here for something like this. It's better just to use plain English

Author: I rewrote this section to use plain english.

parser, symbolic execution engine, etc). Thus, for *N* different languages and
*M* different tools, we have *N x M* different teams and implementations.

> K approaches the problem differently, by imagining how each of the *M*

Reviewer 2: same here

Author: I rewrote this section to use plain english.

     complete proof system;
3.   **Configuration-based:** organize system states into compositional,
     hierarchical, labelled units called cells;
4.   **Rewrite-based:** define system transitions using rewrite rules.

K specifications are compiled into particular *matching logic* theories, giving
them a simple and expressive semantics. K semantic rules are implicitly defined
over the entire configuration structure, but omit unused cells, enabling a
highly modular definitional style. Furthermore, K has been used to develop
programming languages, type systems, and formal analysis tools.

### Manual Objectives

As mentioned in the _Why K?_ section above, the K Framework is designed as a
collection of language-generic command-line interface (CLI) tools which revolve
around K specifications. These tools cover a broad range of uses, but they
typically fall into one of the following categories:

1.  Transforming K Specs (e.g. compilation)
2.  Running K Specs (e.g. concrete and symbolic execution)
3.  Analyzing K Specs (e.g. theorem proving)

The main *user-facing* K tools include:

-   `kompile` - the K compiler driver

> -   `kast` - the stanadlone K parser and abstract syntax tree (AST)

Reviewer 2: let's replace this with `kparse`

Author: Done.

-----------------

Since K specifications are the primary input into the entire system, let us
take a moment to describe them. At the highest level, K specifications describe
a programming language or system using three different pieces:

1.  the *system primitives*, the base datatypes used during system operation,
    e.g., numbers, lists, maps, etc;
2.  the *system state*, a tuple or record over system primitives which gives a
    complete snapshot of the system at any given moment;
3.  the *system behavior*, a set of rules which defines possible system
    evolutions.

K specifications are then defined by a collection of *sentences* which
correspond to the three concepts above:

1.  `syntax` declarations encode the *system primitives*;
2.  `configuration` declarations encode the *system state*;
3.  `context` and `rule` declarations encode the *system behavior*.

K sentences are then organized into one or *modules* which are stored in one or
more *files*. In this scheme, files may *require* other files and modules may
*import* other modules, giving rise to a hierarchy of files and modules. We
give an intuitive sketch of the two levels of grouping in the diagram below:

> ```k

Reviewer 2: this is not k code so the `k` label is not appropriate on this code fence

Author: Removed.

        operators that inhabit our primitive datatypes
    -   **lexical syntax declarations:** define lexemes/tokens for the
        lexer/tokenizer
    -   **syntax associativity declarations:** specify the
        associativity/grouping of our declared operators
    -   **syntax priority declarations:** specify the priority of
        potential ambiguous operators

2.  sentences that define our *system's state*, including:

    -   **configuration declarations:** define labelled, hierarchical records
        using an nested XML-like syntax

3.  sentences that define our *system's behavior*, including:

    -   **context declarations:** describe how primitives and configurations
        can simplify
    -   **context alias declarations:** define templates that can generate new
        contexts
    -   **rule declarations:** define how the system transitions from one state
        to the next

### K Syntax Overview

Here we provide an overview of the top-level K syntax using a K-style BNF
notation. Note that this is not a full definition; we focus instead on the
overall structure of the specification. Syntactic categories which are not

> defined are denoted by names surrounded by curly braces (`{}`). In particular,

Reviewer 2: if we're going to be giving a grammar, we shouldn't make it incomplete like this. It's possible to define the syntax of all of these identifiers and such fairly readily.

Author: I removed this section from this PR because it needs more polish.

we omit the syntax of strings (`{String}`) and all types of identifiers, e.g.,
`{ModuleId}`, `{Id}`, `{SortId}`, `{CellId}`, etc. We also omit the syntax of
attributes (`{Attribute}`), a highly flexible mechanism for modifying the
interpretation of modules and various kinds of sentences.

Due to the highly generic nature of user-defined K syntax, the parsing process
of K specifications roughly proceeds in three stages, corresponding to the
three kinds of declarations shown above.

1.  In the first stage, system primitive declarations are gathered and a parser
    is constructed.

> 2.  In the second stage, the system primitive parser is used to parse

Reviewer 2: better to talk about outer and inner parsing as that is a much more endemic nomenclature to the project.

Author: I removed this section from this PR because it needs more polish.

    configuration declarations; in our BNF grammar below, we use `{TermBubble}`
    to describe things which can be parsed using the primitive syntax.
3.  Finally, a third stage parser is constructed containing both primitive and
    configuration syntax is constructed which is used to parse contexts,
    context aliases, and rules. In our BNF grammar below, we use the words

>     `{Bubble}` and `{BubbleWithRewrites}` to describe parsers which contain the

Reviewer 2: this is incorrect; there is only one type of bubble. What may appear inside that bubble is not part of the outer syntax of K; it's part of a separate grammar that is generated from the user-defined syntax.

Author: I removed this section from this PR because it needs more polish.

    combined primitive and configuration syntax. The `{BubbleWithRewrites}`
    category additional adds the rewrite arrow syntax (`=>`) needed for parsing
    `rule`s and `context`s.

**Note:** _the parsers listed above are all used internally by the K compiler
`kompile`. As part of compilation, `kompile` will generate a separate
language-specific parser that can only parse programs in the specified language
syntax. The language-specific parser is the one invoked when using `krun`._

We now present our high-level BNF syntax:

```
  File         ::= RequiresList ModuleList

  RequiresList ::= "" | Requires RequiresList
  Requires     ::= RequireCmd {String}
  RequireCmd   ::= "requires" | "require"

  Module       ::= "module" {ModuleId} AttributeSet ImportList SentenceList "endmodule"

  AttributeSet ::= "" | "[" Attributes "]"

>   Attributes   ::= Attributes "," {Attribute}

Reviewer 2: this is not correct as your grammar here does not allow for the possibility of a sentence with only a single attribute.

Author: I removed this section from this PR because it needs more polish.

  SentenceList ::= "" | Sentence SentenceList
  Sentence     ::= SortDeclaration
                 | GrammarDeclaration
                 | LexicalDeclaration
                 | AssociativityDeclaration
                 | PriorityDeclaration
                 | ConfigDeclaration
                 | RuleDeclaration
                 | ContextDeclaration
                 | ContextAliasDeclaration
```

We can further drill down by sentence type.

```
  SortDeclaration    ::= "syntax" {SortId} AttributeSet

  GrammarDeclaration ::= "syntax" {SortId} "::=" AlternativeList
  AlternativeList    ::= Production
                       | Production "|" AlternativeList

  Production         ::= ProductionSyntax AttributeSet
  ProductionSyntax   ::= FunctionSyntax
                       | GenericSyntax

  FunctionNotation   ::= {Id} "(" ParameterList ")"
  ParameterList      ::= "" | NeParameterList
  NeParameterList    ::= Parameter | Parameter "," NeParameterList
  Parameter          ::= OptParamName {SortId}
  OptParamName       ::= "" | {Id} ":"

  GenericNotation    ::= StringOrSortIdList
  StringOrSortIdList ::= StringOrSortId | StringOrSortId StringOrSortIdList

>   StringOrSortIdList ::= {String} | {SortId}

Reviewer 2: this is also not correct as you are not allowing for regular expression terminals

Author: I removed this section from this PR because it needs more polish.

  LexicalDeclaration ::= "syntax" "lexical" {Id} "=" {String}

  AssociativityDeclaration ::= "syntax" Associativity NeIdList
  NeIdList                 ::= {Id} | {Id} NeIdList
  Associativity            ::= "left" | "right" | "non-assoc"

>   PriorityDeclaration ::= "syntax" "priorities" PriorityList

Reviewer 2: this can also be written as "syntax priority"

Author: I removed this section from this PR because it needs more polish.

>   PriorityList        ::= NeIdList ">" PriorityList

Reviewer 2: this is also not correct as your grammar does not allow a sentence with only two priority levels

Author: I removed this section from this PR because it needs more polish.

                        | NeIdList ">" NeIdList

>   ConfigDeclaration ::= "configuration" NeCellDeclList

Reviewer 2: this is not really how configuration parsing works.

Author: I removed this section from this PR because it needs more polish.

  NeCellDeclList    ::= CellDecl | CellDecl NeCellDeclList
  CellDecl          ::= ConfigImport
                      | StartCellDecl InnerCellDecl EndCellDecl
  ConfigImport      ::= "<" {CellId} "/>"
  StartCellDecl     ::= "<" {CellId} CellAttributeList ">"
  InnerCellDecl     ::= {TermBubble} | NeCellDeclList
  EndCellDecl       ::= "</" {CellId} ">"
  CellAttributeList ::= "" | CellAttribute CellAttributeList
  CellAttribute     ::= {Id} "=" {String}

  RuleDeclaration   ::= "rule" OptName {BubbleWithRewrites} RequiresCondition EnsuresCondition AttributeSet
  OptName           ::= "" | Name
  Name              ::= "[" {Id} "]" ":"

>   RequiresCondition ::= "" | "requires" {Bubble}

Reviewer 2: this is also not how bubbles in rules work

Author: I removed this section from this PR because it needs more polish.

  EnsuresCondition  ::= "" | "ensures" {Bubble}

  ContextDeclaration      ::= "context" {BubbleWithRewrites} RequiresCondition AttributeSet

>   ContextAliasDeclaration ::= "context" "alias" Name {Bubble} AttributeSet

Reviewer 2: context aliases can have requires clauses.

Author: I removed this section from this PR because it needs more polish.

2.   **Semantic:** correspond to a logical theory with a sound and relatively
     complete proof system;
3.   **Configuration-based:** organize system states into compositional,
     hierarchical, labelled units called cells;
4.   **Rewrite-based:** define system transitions using rewrite rules.

K specifications are compiled into particular *matching logic* theories, giving
them a simple and expressive semantics. K semantic rules are implicitly defined
over the entire configuration structure, but omit unused cells, enabling a
highly modular definitional style. Furthermore, K has been used to develop
programming languages, type systems, and formal analysis tools.

### K Framework Overview

As mentioned in the _Why K?_ section above, the K Framework is designed as a
collection of language-generic command-line interface (CLI) tools which revolve
around K specifications. These tools cover a broad range of uses, but typically
fall into one of the following categories:

1.  Transforming K Specs (e.g. compilation, visual presentation)
2.  Running K Specs (e.g. concrete and symbolic execution)
3.  Analyzing K Specs (e.g. theorem proving, bounded model checking)

> Different tools in the K Framework have different levels of maturity.
> The most well-supported and broadly used tools are the following:

Reviewer: ```suggestion
The main *user-facing* K tools include:
```

> -   `komile` - the K compiler driver

Reviewer: ```suggestion
-   `kompile` - the K compiler driver
```

-   `krun` - the K interpreter and symbolic execution engine driver
-   `kprove` - the K theorem prover
-   `kast` - the stanadlone K parser and abstract syntax tree (AST)
    transformation tool

### Manual Objectives

This user manual is designed to be a tool reference. In particular, it is not
desgined to be a tutorial on how to write K specifications or to teach the
logical foundations of K. New K users should consult our dedicated
[K tutorial](https://kframework.org/k-distribution/k-tutorial/).
Researchers seeking to learn more about the logic underlying K are encouraged
to peruse the
[growing literature on K and matching logic](https://fsl.cs.illinois.edu/projects/pl/index.html).
We will consider the manual complete when it provides a complete description of

> all well-supported K tools and features.

Reviewer: ```suggestion
This user manual is designed to be a tool reference. In particular, it is not
desgined to be a tutorial on how to write K specifications or to teach the
logical foundations of K. New K users should consult our dedicated
[K tutorial](https://kframework.org/k-distribution/k-tutorial/),
or the more language-design oriented [PL tutorial](https://kframework.org/k-distribution/pl-tutorial/).
Researchers seeking to learn more about the logic underlying K are encouraged
to peruse the
[growing literature on K and matching logic](https://fsl.cs.illinois.edu/projects/pl/index.html).
We will consider the manual complete when it provides a complete description of
all well-supported K tools and features.
```
